src/transformers/cache_utils.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABC, abstractmethod
+from typing import Dict, List, Tuple, TypeVar
+import torch
+
+T = TypeVar("T")
+
+
+class Cache(ABC):
+    def __init__(self) -> None:
+        self.cache: Dict[int, Tuple[torch.Tensor]] = {}
+        self.layer_idx = 0
+
+    @abstractmethod
+    def update_pre_rotation(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
+        pass
+
+    @abstractmethod
+    def update(self, key_states, value_states) -> None:
+        pass
+
+    def __getitem__(self, index: int):
+        return self.cache[self.layer_idx][index]
```

Comment:
```suggestion
```

Don't really think we need such a method to begin with no? I'd keep the class as light-weight as possible

Reply:
We do if we want `past_key_values` to remain indexable like it currently is.

Reply:
I was envisioning a bit of a different design here: 
- 1.) We try to avoid any `isinstance(cache, Cache)` statements in all non-public classes such as `LlamaAttention` - instead we assume that all objects the flow inside the model are of the cache class type
- 2.) To keep backwards compatibility, we always convert the DynamicCache to the tuple of tuples (called `past_key_values`) that is expected to be returned from the forward method. Similarly we convert the `past_key_values` to a cache object at the beginning of forward if that isn't the case already. 
I would also go so far as to add a `use_legacy_cache` flag to the forward methods so that we can bit-by-bit deprecate the tuple of tuples `past_key_values`. Also for our `generate` method we could already set this flag to False to not suffer from any potential decrease in speed.

I don't think we can keep 100% backward compatibility between a `cache` object and `past_key_values` anyways so at the moment we cannot return a cache object.

On the other hand, new caches that we implement such as the attention sinks cache of the "compilable" cache should never be returned as a tuple of tuples and also don't have to be backward compatible with a tuple of tuples.

Reply:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8. See this [comment](https://github.com/huggingface/transformers/pull/26681#issuecomment-1759921654) for a follow-up question.